src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java

```diff
@@ -61,17 +64,17 @@ public static void initialize() {
         }
     }
 
-    public static AuthenticationProvider getProvider(String scheme) {
+    public static ServerAuthenticationProvider getProvider(ZooKeeperServer zks, String scheme) {
         if(!initialized)
-            initialize();
-        return authenticationProviders.get(scheme);
+            initialize(zks);
+        return WrappedAuthenticationProvider.wrap(authenticationProviders.get(scheme));
```

Comment:
Differentiating between `ServerAuthenticationProvider` and `AuthenticationProvider` is where the beauty of `WrappedAuthenticationProvider` comes in. `WrappedAuthenticationProvider` does this differentiating. The rest of the code only deals with one API and `WrappedAuthenticationProvider` mitigates the two.